### PR TITLE
Implements type inference for join handles

### PR DESCRIPTION
**Pull request by piotrswigon**
This builds heavily on #4863.

### CODE REVIEW

**Comment by piotrswigon** [comment]
Ping @shans @bgogul

File: src/runtime/recipe/handle.ts [hunk 1]
```diff
@@ -24,6 +24,21 @@ import {ClaimIsTag, Claim} from '../particle-claim.js';
 import {StorageKey} from '../storageNG/storage-key.js';
 import {Capabilities} from '../capabilities.js';
 
+// Represents a relationship between handles.
+// The only example right now is a join handle associating a number
+// of regular handles to expose thier join.
```
[the remark this answers is not in the document]
Comment (piotrswigon) [reply src/runtime/recipe/handle.ts]: Thanks Gogul, fixed!

File: src/runtime/recipe/handle.ts [hunk 2]
```diff
@@ -24,6 +24,26 @@ import {ClaimIsTag, Claim} from '../particle-claim.js';
...
+export class HandleAssociation {
+  constructor(
+    // The handle being associated, e.g. a 'map' handle that is joined.
+    public readonly associate: Handle,
+    // The handle associating other handle, e.g. a 'join' handle exposing a join of data.
+    public readonly associator: Handle,
+    // The type of the association.
+    // 'associate' is being "read" as [T]
+    // 'associator' is being "written to" as [(..., &T, ...)]
+    public readonly type: Type) {}
```
Comment (shans) [inline src/runtime/recipe/handle.ts]: is Type ever divergent from associate's type? If not, then could we avoid complexity by having a list of associated handles directly on the associator?

File: src/runtime/recipe/handle.ts [hunk 3]
```diff
@@ -220,6 +241,8 @@ export class Handle implements Comparable<Handle> {
   get ttl() { return this._ttl; }
   set ttl(ttl: Ttl) { this._ttl = ttl; }
   get isSynthetic() { return this.fate === 'join'; } // Join handles are the first type of synthetic handles, other may come.
+  get associators() { return this._associations.filter(a => a.associate === this); }
+  get associated() { return this._associations.filter(a => a.associator === this); }
```
Comment (shans) [inline src/runtime/recipe/handle.ts]: both directions are hosted in _associations? This makes cloning unlikely to succeed.

Comment (piotrswigon) [reply src/runtime/recipe/handle.ts]: During cloning we do 2 things to ensure it works:
1) We make associator responsible for re-creating these connections (akin to particle being responsible for recreating handle connections)
2) We clone non-synthetic handles first, and synthetic handles second.
There is a test that these connections are recreated correctly after a clone.

It's `it('can normalize and clone a recipe with a synthetic join handle'`

File: src/runtime/recipe/handle.ts (hunk 2 above)
Comment (piotrswigon) [reply src/runtime/recipe/handle.ts]: If associate doesn't have a link to the associator, then when time comes to resolve its type we don't have the constraint that comes from a join.

Have you seen `it('can resolve a recipe with a synthetic join handle and multiple readers'` ? It all works out nicely, multiple particles can read different things from the join as well as mapped handles, and the types get worked out properly by the type system.

Given I have this working properly (I think!) I would prefer not to dumb it down with restrictions that some types have to match if it's not the case in the longer run.

Comment (shans) [reply src/runtime/recipe/handle.ts]: the link can be bidirectional; I'm querying whether we need an additional object.

File: src/runtime/recipe/handle.ts (hunk 3 above)
Comment (shans) [reply src/runtime/recipe/handle.ts]: ah yes, I see that you're using this.associated rather than this._associations. Add a comment explaining cloning strategy above?

Comment (piotrswigon) [reply src/runtime/recipe/handle.ts]: There is a comment explaining the cloning strategy in the recipe.ts, where we first clone regular handles and then synthetic ones. I think that's a better place for it, don't you think?

File: src/runtime/recipe/handle.ts (hunk 2 above)
Comment (piotrswigon) [reply src/runtime/recipe/handle.ts]: Ok, you were correct. I did manage to get this to work without the additional type variable (which was the only thing that was hanging off this relationship object) and having the same type being the type of a handle and the type exposed in the join.

Good job pushing me to simplify this.

File: src/runtime/recipe/handle.ts (hunk 3 above)
Comment (shans) [reply src/runtime/recipe/handle.ts]: yup, that makes sense to me.

File: src/runtime/recipe/handle.ts (hunk 2 above)
Comment (shans) [reply src/runtime/recipe/handle.ts]: Thanks for bearing with me! This looks super neat now.
